regex-capi/src/lib.rs

```diff
@@ -0,0 +1,515 @@
+extern crate libc;
+extern crate regex;
+
+use std::collections::HashMap;
+use std::ffi::{CStr, CString};
+use std::fmt;
+use std::io::{self, Write};
+use std::ops::Deref;
+use std::panic::{self, AssertUnwindSafe};
+use std::process;
+use std::ptr;
+use std::slice;
+use std::str;
+
+use libc::{c_char, uint32_t};
+use regex::bytes;
+
+macro_rules! protect {
+    ($body:block) => {
+        match panic::catch_unwind(AssertUnwindSafe(move || $body)) {
+            Ok(v) => v,
+            Err(err) => {
+                let msg = if let Some(&s) = err.downcast_ref::<&str>() {
+                    s.to_owned()
+                } else if let Some(s) = err.downcast_ref::<String>() {
+                    s.to_owned()
+                } else {
+                    "UNABLE TO SHOW RESULT OF PANIC.".to_owned()
+                };
+                let _ = writeln!(
+                    &mut io::stderr(),
+                    "panic unwind caught, aborting: {:?}",
+                    msg);
+                process::exit(1)
```

Comment:
Interesting. I didn't think at all about the distinction between `process::exit(...)` and `abort`. Happy to switch!